### PR TITLE
refactor state crate

### LINKED ISSUE #2486: state crate refactor
- unify names in the `namada_state` as its usages (`struct State` was previously `Storage`, but we still have e.g. `WlStorage` that wraps the `State`, `WriteLogAndStorage` [and its methods], `TempWlStorage`, inside `Shell`, etc.)
- make the `State` internal to the crate, from the outside it should only be accessed via what's currently called `WlStorage` - again, the names are confusing)

### PR DESCRIPTION
**Pull request by tzemanovic**
## Describe your changes

closes #2486

### High-level

- Renamed `struct State` to `struct InMemory` and removed its field `db: D` together with `D` type param
  - Most methods on `State` that were using the `db` are moved to `trait State` and prefixed with `db_` (e.g. `State::read` becomes `State::db_read`)
- Removed `struct WlStorage`, `struct TempWlStorage` and `trait WriteLogAndStorage`
- Added `trait State` generic over DB and hasher, comprised of a write log, DB and `InMemory` state.
- Added `struct FullAccessState` (impls `State`, `StorageRead` and `StorageWrite`) - this is owned by shell which has full access to everything (DB, write log and `InMemory` state)
- Added `struct WlState` (impls `State`, `StorageRead` and `StorageWrite`) - DB is read only, write/delete can only be done via write log
- Added `struct TempWlState` (impls `State`, `StorageRead` and `StorageWrite`) - replaces `TempWlStorage` - for prepare and process proposal and dry-run txs which use a temporary write log
- Added `struct VpHostEnvState` (impls `State` and `StorageRead`) - state for VP exec, everything is read-only (except for gas meter which is held in a `RefCell` to allow mutation in immutable methods)
- Added `struct TxHostEnvState` (impls `State`, `StorageRead` and `StorageWrite`) - state for tx exec, can only write to write log, cannot modify in-memory state. Includes gas accounting (gas meter also in a `RefCell`).
- ~~`struct TxCtx` now hold reference to a generic mutable `State`~~
  - I couldn't find a way to make this work because the type has to have a `'static` lifetime so anything that contains a non-static reference is a no-go. Instead, the `TxCtx` now has a `state()` method, which returns a `TxHostEnvState` that allows to use `StorageRead + StorageWrite` in host env with gas accounting included (replaces temporary implementation we've added for IBC tx workaround)
  - same deal with host_env fns, they all have be `'static` to be able to use them in `wasm_imports`
- ~~`struct VpCtx` and `native_vp::Ctx` now hold a reference to a generic immutable `State`~~
  - Same as above, `VpCtx` has a `state()` method that returns `VpHostEnvState`
- Replaced `struct TestWlStorage` by `type TestState = FullAccessState<MockDB, Sha256Hasher>`

### Usage

- From shell, we can restrict state access with:
  - `FullAccessState::restrict_writes_to_write_log` - returns `&mut WlState` for transactions and other state changes in a block
  - `FullAccessState::read_only` - returns `&WlState` for VPs and queries
- `trait State` components access
  - `State::write_log()`, `State::write_log_mut()`, `State::db()`, `State::in_mem()` (`InMemory` state)
  - `State::split_borrow` to use all 3 at the same time
- Most things benefit from using `trait State` in place of `WlStorage` to simplify the verbose bounds:
  ```rust
      DB: namada_state::DB + for<'iter> namada_state::DBIter<'iter>,
      H: StorageHasher,
  ```
  becomes
  ```rust
      S: State
  ```

### Next steps

For sub-system modules envisioned in https://github.com/anoma/namada/issues/2111 we'll want to replace the usage of `WlStorage`/`WlState`/`State` with the more abstract `StorageRead + StorageWrite` with additional traits that add access to individual parts of the state that are being used but not covered by `StorageRead` or `StorageWrite` (e.g. things from `InMemory` state - for shielded token we have `trait WithConversionState` that provides mutable access to conversion state (in the concrete `FullAccessState` this is held in `struct InMemory`). The only places that directly use the state will be:
  - The shell that owns the state
  - High-level ABCI handlers
  - Tx and VP environments (but not tx and VP implementations)
These are `governance`, `ethereum_bridge` and `ibc`.

## TODOs
- [x] make it build
- [x] replace the temporary `StorageRead` and `StorageWrite` impl for `TxCtx` in `crates/namada/src/vm/host_env.rs` with a impl provided by the state crate (must account for gas)
- [x] fix tests

## Indicate on which release or other PRs this topic is based on

#2507 - diff https://github.com/anoma/namada/pull/2606/files/d44bf1e2c242e076ca96e23c1e27dfba663c12b2..daffb53860531bf589f00259c3e12d4e5ee3b395

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

**Comment by tzemanovic** [comment]
there's an issue with loading merkle tree after node restart, looking into it

File: crates/apps/src/lib/node/ledger/shell/finalize_block.rs [hunk 1]
```diff
@@ -1932,41 +1908,35 @@ mod test_finalize_block {
...
         let current_height = 4_u64;
-        assert_eq!(current_height, shell.wl_storage.storage.block.height.0);
+        assert_eq!(current_height, shell.state.in_mem().block.height.0);
 
         for _ in current_height..height_of_next_epoch.0 + 2 {
-            dbg!(
-                get_rewards_acc(&shell.wl_storage),
-                get_rewards_sum(&shell.wl_storage),
-            );
+            // dbg!(get_rewards_acc(&shell.state),
```
Comment (grarco) [inline crates/apps/src/lib/node/ledger/shell/finalize_block.rs:1916]: Don't we need this dbg anymore?

File: crates/apps/src/lib/node/ledger/shell/finalize_block.rs [hunk 2]
```diff
@@ -4191,7 +4124,7 @@ mod test_finalize_block {
             Dec::one(),
             Dec::new(9, 0).unwrap() * tot_frac * tot_frac,
         );
-        dbg!(&cubic_rate);
+        // dbg!(&cubic_rate);
```
Comment (grarco) [inline crates/apps/src/lib/node/ledger/shell/finalize_block.rs:4127]: Same here

File: crates/ibc/src/actions.rs [hunk 3]
```diff
@@ -4,77 +4,244 @@ use std::cell::RefCell;
...
+    ) -> StorageResult<()> {
+        self.state.write_bytes(key, val)
+    }
+
+    fn delete(&mut self, key: &namada_storage::Key) -> StorageResult<()> {
+        self.state.delete(key)
+    }
+}
+
+// impl<S> StateRead for IbcProtocolContext<'_, S>
```
Comment (grarco) [inline crates/ibc/src/actions.rs:119]: I think we can remove this commented block if not needed

File: crates/namada/src/ledger/mod.rs [hunk 4]
```diff
@@ -184,25 +185,22 @@ mod test {
...
             // Initialize mock gas limit
             let max_block_gas_key =
                 namada_parameters::storage::get_max_block_gas_key();
-            wl_storage
-                .storage
-                .write(&max_block_gas_key, namada_core::encode(&20_000_000_u64))
-                .expect(
-                    "Max block gas parameter must be initialized in storage",
-                );
+            state.write(&max_block_gas_key, 20_000_000_u64).expect(
```
Comment (grarco) [inline crates/namada/src/ledger/mod.rs:193]: Shouldn't this be a `db_write`? Probably it doesn't really affect the tests though but I saw that previously we were writing directly t storage while now it looks like we are writing to the block write log

**Comment by tzemanovic** [comment]
> there's an issue with loading merkle tree after node restart, looking into it

fixed in [1519e71](https://github.com/anoma/namada/pull/2606/commits/1519e714897f95ff2373e4c31e588d6f5370cc27)

File: crates/namada/src/ledger/vp_host_fns.rs [hunk 5]
```diff
@@ -50,9 +52,9 @@ pub type EnvResult<T> = std::result::Result<T, RuntimeError>;
 
 /// Add a gas cost incured in a validity predicate
 pub fn add_gas(
-    gas_meter: &mut VpGasMeter,
+    gas_meter: &mut impl DerefMut<Target = VpGasMeter>,
```
Comment (grarco) [inline crates/namada/src/ledger/vp_host_fns.rs:55]: Do we need the `&mut `?

Comment (grarco) [inline crates/namada/src/ledger/vp_host_fns.rs:55]: Do we need the `&mut` when requesting `DerefMut`?

File: crates/namada/src/vm/host_env.rs [hunk 6]
```diff
@@ -815,94 +835,97 @@ where
...
     }
 
-    check_address_existence(env, &key)?;
+    check_address_existence::<MEM, D, H, CA>(env, &key)?;
 
-    let write_log = unsafe { env.ctx.write_log.get() };
-    let (gas, _size_diff) = write_log
-        .write(&key, value)
-        .map_err(TxRuntimeError::StorageModificationError)?;
-    tx_charge_gas(env, gas)
```
Comment (grarco) [inline crates/namada/src/vm/host_env.rs:855]: I think we lost this gas

File: crates/sdk/src/queries/mod.rs [hunk 7]
```diff
@@ -126,21 +127,18 @@ mod testing {
...
             // Initialize mock gas limit
             let max_block_gas_key =
                 namada_parameters::storage::get_max_block_gas_key();
-            wl_storage
-                .storage
-                .write(&max_block_gas_key, namada_core::encode(&20_000_000_u64))
-                .expect(
-                    "Max block gas parameter must be initialized in storage",
-                );
+            state.write(&max_block_gas_key, 20_000_000_u64).expect(
```
Comment (grarco) [inline crates/sdk/src/queries/mod.rs:135]: `db_write` here?

File: crates/apps/src/lib/node/ledger/shell/finalize_block.rs (hunk 1 above)
Comment (tzemanovic) [reply crates/apps/src/lib/node/ledger/shell/finalize_block.rs:1916]: we shouldn't need these unless the tests are failing but I'll put them back to leave them as they are

File: crates/namada/src/ledger/mod.rs (hunk 4 above)
Comment (tzemanovic) [reply crates/namada/src/ledger/mod.rs:193]: yeah, the storage read will still find it, but it really should go directly to DB

File: crates/namada/src/ledger/protocol/mod.rs [hunk 8]
```diff
@@ -564,37 +567,30 @@ where
...
-        let (write_log, storage) = wl_storage.split_borrow();
-        (
-            tx_gas_meter,
-            storage,
-            write_log,
-            vp_wasm_cache,
-            tx_wasm_cache,
-        )
-    };
+    let (tx_gas_meter, vp_wasm_cache, tx_wasm_cache) =
```
[the remark this answers is not in the document]
Comment (tzemanovic) [reply crates/namada/src/ledger/protocol/mod.rs:588]: ah yes

File: crates/namada/src/ledger/vp_host_fns.rs (hunk 5 above)
Comment (tzemanovic) [reply crates/namada/src/ledger/vp_host_fns.rs:55]: Good q - we have the same `RefCell` type in native and wasm env, so I'll change these to just pass it around

File: crates/apps/src/lib/node/ledger/shell/finalize_block.rs (hunk 1 above)
Comment (tzemanovic) [reply crates/apps/src/lib/node/ledger/shell/finalize_block.rs]: done in 43bc13e61b40c67fb35b552e170b2f77d93918c6

File: crates/apps/src/lib/node/ledger/shell/finalize_block.rs (hunk 2 above)
Comment (tzemanovic) [reply crates/apps/src/lib/node/ledger/shell/finalize_block.rs]: 43bc13e61b40c67fb35b552e170b2f77d93918c6

File: crates/ibc/src/actions.rs (hunk 3 above)
Comment (tzemanovic) [reply crates/ibc/src/actions.rs]: done in c8f43714b0381ea0ac091a09dade646d69109135

File: crates/namada/src/ledger/mod.rs (hunk 4 above)
Comment (tzemanovic) [reply crates/namada/src/ledger/mod.rs]: c3c29e5ec22d3a1ae6e0f554b7115297fb25c532

File: crates/namada/src/ledger/protocol/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tzemanovic) [reply crates/namada/src/ledger/protocol/mod.rs]: c3c29e5ec22d3a1ae6e0f554b7115297fb25c532

File: crates/sdk/src/queries/mod.rs (hunk 7 above)
Comment (tzemanovic) [reply crates/sdk/src/queries/mod.rs]: b663ac00ae63522dcb9b648aabb806d2acc0929a

File: crates/namada/src/ledger/vp_host_fns.rs (hunk 5 above)
Comment (tzemanovic) [reply crates/namada/src/ledger/vp_host_fns.rs]: done in 0c98d6d7d30d79ef959ebfa6ce020b9053122fc5

**Comment by tzemanovic** [comment]
squashed fixups (no diff https://github.com/anoma/namada/compare/0c98d6d7d30d79ef959ebfa6ce020b9053122fc5..daffb53860531bf589f00259c3e12d4e5ee3b395)

**Pull request merged**